src/ripple/conditions/impl/Condition.cpp

```diff
@@ -0,0 +1,180 @@
+#include <ripple/conditions/Fulfillment.h>
+#include <ripple/conditions/impl/utils.h>
+#include <boost/regex.hpp>
+#include <boost/optional.hpp>
+#include <vector>
+#include <iostream>
+
+namespace ripple {
+namespace cryptoconditions {
+
+boost::optional<Condition>
+loadCondition(std::string const& s)
+{
+    static boost::regex const re_current (
+            "^"                            // start of line
+            "cc:"                          // 'cc' for cryptocondition
+            "([1-9a-f][0-9a-f]{0,3}|0):"   // type (hexadecimal)
+            "([1-9a-f][0-9a-f]{0,15}):"    // feature bitmask (hexadecimal)
+            "([a-zA-Z0-9_-]{0,86}):"       // fingerprint (base64url)
+            "([1-9][0-9]{0,17}|0)"         // fulfillment length (decimal)
+            "$"                            // end of line
+        , boost::regex_constants::optimize
+    );
+
+    boost::smatch match;
+
+    if (!boost::regex_match (s, match, re_current))
+        return boost::none;
+
+    try
+    {
+        Condition c;
+
+        c.type = parse_hexadecimal<std::uint16_t> (match[1]);
+
+        if (!isCondition (c.type))
+            return boost::none;
+
+        c.featureBitmask = parse_hexadecimal<std::uint32_t>(match[2]);
+        c.maxFulfillmentLength = parse_decimal<std::uint16_t>(match[4]);
+
+        if (c.maxFulfillmentLength > maxSupportedFulfillmentLength)
+            return boost::none;
+
+        // TODO: Avoid copying by decoding directly
+        // into the condition's buffer
+        auto fingerprint = base64url_decode(match[3]);
+
+        if (fingerprint.size() != c.fingerprint.size())
+            return boost::none;
+
+        std::memcpy(
+            c.fingerprint.data(),
+            fingerprint.data(),
+            fingerprint.size());
+
+        return c;
+    }
+    catch (std::exception const& x)
+    {
+        return boost::none;
+    }
+}
+
+boost::optional<Condition>
+loadCondition(Slice s)
+{
+    try
+    {
+        auto iter = s.cbegin();
+
+        Condition c;
+
+        std::tie (iter, c.type) =
+            oer::decode_integer<std::uint16_t> (
+                iter, s.cend());
+
+        if (!isCondition (c.type))
+            return boost::none;
+
+        std::tie (iter, c.featureBitmask) =
+            oer::decode_varuint<std::uint32_t> (
+                iter, s.cend());
+
+        {
+            std::size_t len;
+
+            std::tie (iter, len) = oer::decode_length (iter, s.cend());
+
+            // Incorrect signature length
+            if (len != c.fingerprint.size())
+                return boost::none;
+
+            // Short buffer
+            if (std::distance (iter, s.cend()) < len)
+                return boost::none;
+
+            auto p = c.fingerprint.data();
+
+            while (len--)
+                *p++ = *iter++;
+        }
+
+        std::tie (iter, c.maxFulfillmentLength) =
+            oer::decode_varuint<std::uint16_t> (
+                iter, s.cend());
```

Comment:
If `iter==s.cend` here (which can sill pass the short buffer test above) then `decode_varuint` will dereference `*iter` when it calls `decode_length`, reading memory beyond the buffer.